core/src/main/scala/kafka/log/ProducerStateManager.scala

```diff
@@ -107,13 +107,18 @@ private[log] class ProducerAppendInfo(val producerId: Long,
         throw new DuplicateSequenceNumberException(s"Duplicate sequence number for producerId $producerId: (incomingBatch.firstSeq, " +
           s"incomingBatch.lastSeq): ($firstSeq, $lastSeq), (lastEntry.firstSeq, lastEntry.lastSeq): " +
           s"(${this.firstSeq}, ${this.lastSeq}).")
-      } else if (firstSeq != this.lastSeq + 1L) {
+      } else if (firstSeq != this.lastSeq + 1L && (firstSeq != 0 && this.lastSeq != Short.MaxValue)) {
```

Comment:
Actually I think @guozhangwang is right. Let's right this as `!((firstSeq == lastSeq + 1) || (firstSeq == 0 && lastSeq == max))` because I think we agree it's clearer.